L1Trigger/Phase2L1ParticleFlow/src/BitwisePFAlgo.cc

```diff
@@ -0,0 +1,204 @@
+#include "L1Trigger/Phase2L1ParticleFlow/interface/BitwisePFAlgo.h"
+#include "FWCore/Utilities/interface/Exception.h"
+
+//#define REG_HGCal
+#include "ref/pfalgo2hgc_ref.h"
+#include "ref/pfalgo3_ref.h"
+#include "utils/DiscretePF2Firmware.h"
+#include "utils/Firmware2DiscretePF.h"
+
+using namespace l1tpf_impl;
+
+BitwisePFAlgo::BitwisePFAlgo(const edm::ParameterSet &iConfig) : PFAlgoBase(iConfig), config_(nullptr) {
+  const edm::ParameterSet &bitwiseConfig = iConfig.getParameter<edm::ParameterSet>("bitwiseConfig");
+  const std::string &algo = iConfig.getParameter<std::string>("bitwiseAlgo");
+  debug_ = iConfig.getUntrackedParameter<int>("debugBitwisePFAlgo", iConfig.getUntrackedParameter<int>("debug", 0));
+  if (algo == "pfalgo3") {
+    algo_ = algo3;
+    config_ = new pfalgo3_config(bitwiseConfig.getParameter<uint32_t>("NTRACK"),
```

Comment:
use smart pointer

Reply:
unique_ptr wouldn't work because in the header file I have a forward declaration (to avoid inlcuding all the firmware headers in the rest of the code) and it wouldn't know which deleter to use.
I can make it a shared_ptr if it's preferable (though it has some overhead)

Reply:
The class definition is complete when it's constructed, though, right? It's not clear to me why unique_ptr wouldn't work in that case. (If it's an issue with the class destructor, just define the destructor in the .cc file)

Reply:
The base class' destructor is not `virtual` and therefore deleting a deriving class via pointer to base class would lead to wrong destructor to be called.

Reply:
In that case, shouldn't the base class be fixed?

Reply:
this is done in https://github.com/gpetruc/cmssw/commit/ccbc0d8048dd43d7b79671d57b7bbcb06ac0a4bf#diff-f3806d050f9dc0c13d3e5d9e616a4b72 (will be included by Vladimir when updating the PR)